projects/vulkan-loader/fuzzers/instance_create_fuzzer.c

```diff
@@ -0,0 +1,119 @@
+/* Copyright 2023 Google LLC
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "cJSON.h"
+#include "loader.h"
+
+/*
+ * Create config files for given path and data.
+ */
+int create_config_file(const char* config_path, const char* config_filename, const uint8_t* data, size_t size) {
+  char filename[512];
+  char path[256];
+  char command[256];
+
+  sprintf(path, "%s/%s", getenv("HOME"), config_path);
+  sprintf(command, "mkdir -p %s", path);
+
+  system(command);
+
+  sprintf(filename, "%s/%s", path, config_filename);
+
+  FILE *fp = fopen(filename, "wb");
+  if (!fp) {
+    return 1;
+  }
+  fwrite(data, size, 1, fp);
+  fclose(fp);
+
+  return 0;
+}
+
+/*
+ * Remove config file
+ */
+void remove_config_file(const char* config_path, const char* config_filename) {
+  char filename[512];
+  sprintf(filename, "%s/%s/%s", getenv("HOME"), config_path, config_filename);
+  unlink(filename);
+}
+
+/*
+ * Targets the instance creation.
+ */
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+  setenv("HOME", "/tmp", 1);
+
+  // Create implicit layer configuration file
+  int result = create_config_file(".local/share/vulkan/implicit_layer.d", "complex_layer.json", data, size);
+  if (result) {
+    return 0;
+  }
+  
+  // Create loader configuration file
+  result = create_config_file(".local/share/vulkan/loader_settings.d", "vk_loader_settings.json", data, size);
+  if (result) {
+    return 0;
+  }
+
+  // Create icd configuration file
+  result = create_config_file(".local/share/vulkan/icd.d", "icd_test.json", data, size);
```

Comment:
In the three cases here, the same data is used for each configuration file, as such, they will all be forced to be the same when `vkCreateInstance` is called. This will likely have a constraint on what can be run, if `vkCreateInstance` depends on the configuration files and there is no explicit requirement for them to be equal.

Could you ensure each part of the buffer from the fuzzer is only used in one place throughout the fuzzers (unless there is a specific dependency that some data should be equal to other data, or, there is no use of the data in the same context).

Reply:
There are three different configs, but they can be merged into one file. This is because all the configs have their own keys, and these keys are not the same. And the fuzzer buffer would only go to one place at a time.

Reply:
I don't follow your logic: the files have different keys so in order to generate valid configuration files they each need to have different content? If so, then using the same buffer of data to write to the configuration files guarantees this never occurs, thus, we should be sure to separate the content?

Reply:
I can split configs, just to make it easier I have merged them. They don't intersect keys, and if json format would be broken overall vulkan-loader would finish parsing and exit from initialization, as it would be if any of configs would be broken. So it doesn't look bad to me, if you want to be more sure about configs, I can split them and create more fuzzers to handle configs separately. So the first fuzzer for initialization would fuzz only layer config, and the next fuzzer for initialization would fuzz only icd config, etc.

Reply:
>  They don't intersect keys,

That's the reason why we should separate the data? Currently all the files will be equal because you use the same data for each file. Thus, there will currently not be a situation where they have different keys -- they will all be equal.

Reply:
> They don't intersect keys,
That a reason why we are able to not split configurations, and use them in one file. And yes it's okay to store them all in one file, with same data.